WalletWasabi.Tests/UnitTests/Blockchain/TransactionBuilding/BranchAndBoundTests.cs

```diff
@@ -13,81 +14,162 @@ public class BranchAndBoundTests
 
 		Assert.False(wasSuccessful);
 		Assert.Null(selectedCoins);
 	}
 
 	[Fact]
-	public void ReturnNullIfNoExactMatchFoundTest()
+	public void ExactMatch_ReturnNullIfNoExactMatchFoundTest()
 	{
 		List<long> inputValues = new() { 120_000, 100_000, 100_000, 50_000, 40_000 };
-		BranchAndBound selector = new(inputValues);
 		long target = 300000;
 
-		bool wasSuccessful = selector.TryGetExactMatch(target, out List<long>? selectedCoins);
+		BranchAndBound algorithm = new(inputValues);
+		ExactMatchStrategy strategy = new(target);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+	}
+
+	/// <summary>Tests that a best found selection is found when an exact solution does not exist.</summary>
+	[Fact]
+	public void CheapestSelection_NoInputCosts()
+	{
+		List<long> inputValues = new() { 11, 7, 5, 3, 2 }; // Sum is 28.
+		long[] inputCosts = new long[] { 0, 0, 0, 0, 0 }; // No input costs. Idealized.
+		long target = 27; // Target that we cannot get as a sum of input values.
+
+		BranchAndBound algorithm = new(inputValues);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
 
 		Assert.False(wasSuccessful);
 		Assert.Null(selectedCoins);
+		Assert.Equal(new long[] { 11, 7, 5, 3, 2 }, strategy.GetBestSelectionFound());
+	}
+
+	[Fact]
+	public void CheapestSelection_ExactMatchIsAlsoCheapest()
+	{
+		List<long> inputValues = new() { 35, 17, 10, 5, 3, 2 };
+
+		// Make the second input very expensive to spend so that it is not selected (not likely in reality).
+		long[] inputCosts = new long[] { 1, 5, 1, 1, 1, 1 };
+
+		long target = 27;
+
+		BranchAndBound algorithm = new(inputValues);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+
+		Assert.Equal(new long[] { 17, 10 }, strategy.GetBestSelectionFound());
+	}
+
+	[Fact]
+	public void CheapestSelection_ExactMatchIsExpensive()
+	{
+		List<long> inputValues = new() { 35, 17, 10, 5, 3, 2 };
+
+		// Make the second input very expensive to spend so that it is not selected (not likely in reality).
+		long[] inputCosts = new long[] { 1, 1000, 1, 1, 1, 1 };
+
+		long target = 27; // Target that we cannot get as a sum of input values.
+
+		BranchAndBound algorithm = new(inputValues);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+
+		// Selection (17, 10) is actually more expensive: 17 + 10 + 1000 + 1 = 1018.
+		// Whereas (35) costs us 35 + 1 = 36.
+		Assert.Equal(new long[] { 35 }, strategy.GetBestSelectionFound());
```

Comment:
Same here, `strategy.GetBestSelectionFound()` can give back different results.
For example:
`Int64[] [ 17, 5, 3, 2 ]`
`Int64[] [ 35 ]`

Reply:
Hopefully, I have fixed this.